storage/src/store/mod.rs

```diff
@@ -0,0 +1,822 @@
+    V: Span,
+    T: Translator,
+{
+    /// Initializes a new [`Store`] database with the given configuration.
+    ///
+    /// ## Rollback
+    ///
+    /// Any uncommitted operations will be rolled back if the [`Store`] was previously closed without
+    /// committing.
+    pub async fn init(
+        context: E,
+        cfg: Config<T, <Operation<K, V> as Read>::Cfg>,
+    ) -> Result<Self, Error> {
+        let mut snapshot: Index<T, u64> =
+            Index::init(context.with_label("snapshot"), cfg.translator);
+
+        let mut log = VJournal::init(
+            context.with_label("log"),
+            VConfig {
+                partition: cfg.log_journal_partition,
+                compression: cfg.log_compression,
+                codec_config: cfg.log_codec_config,
+                write_buffer: cfg.log_write_buffer,
+            },
+        )
+        .await?;
+
+        let mut locations = FJournal::init(
+            context.with_label("locations"),
+            FConfig {
+                partition: cfg.locations_journal_partition,
+                items_per_blob: cfg.locations_items_per_blob,
+                write_buffer: cfg.log_write_buffer,
+                buffer_pool: cfg.buffer_pool.clone(),
+            },
+        )
+        .await?;
+
+        // Build the snapshot by replaying the log of operations.
+        let (op_count, oldest_retained_loc, inactivity_floor_loc) = Self::build_snapshot_from_log(
+            &mut log,
+            &mut locations,
+            &mut snapshot,
+            cfg.log_items_per_section,
+        )
+        .await?;
+
+        Ok(Self {
+            log,
+            snapshot,
+            log_items_per_section: cfg.log_items_per_section,
+            locations,
+            inactivity_floor_loc,
+            oldest_retained_loc,
+            op_count,
+            uncommitted_ops: 0,
+        })
+    }
+
+    /// Gets the value associated with the given key in the store.
+    ///
+    /// If the key does not exist, returns `Ok(None)`.
+    pub async fn get(&self, key: &K) -> Result<Option<V>, Error> {
+        for location in self.snapshot.get(key) {
+            let (k, v) = Self::get_set_op(
+                &self.log,
+                &self.locations,
+                self.log_items_per_section,
+                *location,
+            )
+            .await?;
+
+            if &k == key {
+                return Ok(Some(v));
+            }
+        }
+
+        Ok(None)
+    }
+
+    /// Updates the value associated with the given key in the store.
+    ///
+    /// If the key already has the same value, this is a no-op and no log entry is created.
+    ///
+    /// The operation is immediately visible in the snapshot for subsequent queries, but remains
+    /// uncommitted until [`Store::commit`] is called. Uncommitted operations will be rolled back
+    /// if the store is closed without committing.
+    pub async fn update(&mut self, key: K, value: V) -> Result<UpdateResult, Error> {
```

Comment:
On mobile and haven't taken a full look (so ignore if wrong) but I think you're missing a delete op?

Reply:
Yeah - the `Store` currently doesn't support deletions, only overriding existing keys. Would you like that feature as well here?

https://github.com/commonwarexyz/monorepo/blob/aa42c41252d2348e509ff16b826058222a6a9035/storage/src/store/mod.rs#L5-L7